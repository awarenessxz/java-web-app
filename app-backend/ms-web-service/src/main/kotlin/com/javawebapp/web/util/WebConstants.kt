package com.javawebapp.web.util

object WebConstants {
    // Types of User Roles
    const val USER_ROLE_NORMAL = "NORMAL_USER"
    const val USER_ROLE_ADMIN = "ADMIN_USER"
}