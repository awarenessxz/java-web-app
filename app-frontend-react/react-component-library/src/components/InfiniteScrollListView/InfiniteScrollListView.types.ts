export interface ListViewItem {
    id: string;
    title: string;
    message: string;
    displayDateTime: string;
    isFlagged?: string | boolean;
    originalData?: any;
}

export type DataToItemMapping = ListViewItem;

export interface InfiniteScrollListViewProps {
    /** Data to Item Mapping (Refer to Documentation) */
    dataToItemMapping: DataToItemMapping;
    /** Data API URL eg. /api/something (Refer to Documentation) */
    dataApiUrl: string;
    /** Amount of data to fetch per load -- refer to sql LIMIT or Spring Pagination PAGE_SIZE */
    dataLimit: number;
    /** data fetch position -- refer to sql OFFSET or Spring Pagination PAGE_NO */
    dataOffset: number;
    /** data offset increment amount -- refer to sql OFFSET increment or Spring Pagination PAGE_NO increment */
    dataOffsetIncrement: number;
    /** [OPTIONAL] list of data id which indicates if isFlag should be disabled (typically for read/unread behaviour) */
    dataIsFlaggedReadList?: string[];
    /** message for user when there is no data retrieved from api */
    noDataMsg?: string;
    /** handler for when item in list is clicked */
    onItemClick: (item: any) => void;
    /** height */
    height: string;
    /** width */
    width?: string;
}

export interface InfiniteScrollListViewState {
    dataOffset: number; // refer to sql OFFSET
    isLoading: boolean; // loading new data
    hasNewData: boolean; // new data loaded -- determines if there are more data
    error: string | boolean; // loading error
    data: ListViewItem[];
    triggerScrollRegister: boolean; // toggle to trigger useEffect to register scroll event
}
