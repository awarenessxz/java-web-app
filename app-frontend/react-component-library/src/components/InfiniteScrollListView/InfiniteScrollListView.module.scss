.islvRoot {
  overflow-y: scroll;
  width: 100%;
  max-width: 36ch;
  background-color: #fafbfd;
  text-align: center;
  border: 1px solid #000000;
  padding: 1px;
}

.islvIsFlagged {
  border-left: 6px solid #a64452;
}

.islvIsNotFlagged {
  border-left: 6px solid transparent;
}