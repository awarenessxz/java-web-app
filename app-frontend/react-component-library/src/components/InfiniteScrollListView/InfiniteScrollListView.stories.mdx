import { Meta, Story, ArgsTable } from '@storybook/addon-docs/blocks';
import InfiniteScrollListView from "./InfiniteScrollListView";

<Meta title='Docs/InfiniteScrollListView' component={InfiniteScrollListView} />

# Infinite Scrolling List View Component

This is a infinite scrolling list view that is meant to mimic how outlook email list. This project is made possible
because of [joshtronic](https://www.digitalocean.com/community/tutorials/react-react-infinite-scroll). Check out his
explanation on how to implement infinite scrolling in React Component.

**Future Improvement**
- currently new data is appended into the list inside the component. If there is a change in order, this will break the
component. Hence, improvement will have to be made on data that changes dynamically.
- Read/Unread Feature
- Badges
- Write Test Cases!!

## Usage

To use this component, do the following:

1. Add the library as dependency
    - `yarn add react-component-library`
2. Import the component
    ```
    import { InfiniteScrollListView } from 'react-component-library';
    ```

## Properties

<ArgsTable of={InfiniteScrollListView} />

### Data API

The component will be dynamically calling a rest api to provide it with data. Below are some of the format you should adhere to:

#### 1. API end point

You should expose an API end point in the following format `/any/api/url?limit=?&offset=?`. The component assumes that
the end point is in the format above. Hence, a clearer picture of what the props in the properties table will be as
follows:
- dataApiUrl -- api url `eg. /any/api/url`
- dataLimit -- amount of data to fetch (Refer to SQL)
- dataOffset -- current position of data (Refer to SQL)

#### 2. Data to Item Mapping

This components accepts data retrieved from API in the following format:

```
{
    id: string;
    title: string;
    message: string;
    displayDateTime: string;        # obtained via moment(). Eg. 2020-12-08T04:53:51.822Z
    isFlagged: boolean              # [Optional] -- default = false
}
```

Hence, we provided an `dataToItemMapping` prop for you to map the data which your api returns to the format accepted by
the component. Below is an example scenario.

- Fetch API returns the following object
    ```
    {
        id: '123123'                               # (string)
        header: 'something',                       # (string)
        result: 'asdasdasd',                       # (string)
        creationDate: '2020-12-08T04:53:51.822Z',  # obtained via moment() -- (string)
        isActive: true                             # indicator whether to show a flag -- (boolean)
    }
    ```

- When using the Component, set the `dataToItemMapping` parameter as shown
    ```
    <InfiniteScrollListView
        ...
        dataToItemMapping: {
            id: 'id',                       # maps id to id (string)
            title: 'header',                # maps header to title (string)
            message: 'result',              # maps result to message (string)
            displayDateTime: 'creationDate' # maps creationDate to displayDateTime (string)
            isFlagged: 'isActive'           # maps isActive to isFlagged (boolean)
        },
        ...
    />

#### 3. onItemClick callback

OnItemClick callback will be called when the an item in the scroll list is clicked. The original data which was retrieved
from the rest API would be returned. Important things to note:

- For props `dataIsFlaggedReadList`, the component will keep a state of the "id" inside `dataIsFlaggedReadList`. Whenever
there are any item clicks, the component will update the **state inside the component** and **NOT the list** which was
passed into the component via `dataIsFlaggedReadList`. Hence, if you want to update the list which you pass into the props,
please handle the logic on your own.
