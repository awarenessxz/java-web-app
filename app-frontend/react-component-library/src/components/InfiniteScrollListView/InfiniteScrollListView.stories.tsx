import * as React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import InfiniteScrollListView from './InfiniteScrollListView';
import { InfiniteScrollListViewProps, ListViewItem } from './InfiniteScrollListView.types';

export default {
    title: 'Components/InfiniteScrollListView',
    component: InfiniteScrollListView,
    excludeStories: /.*Data$/,
} as Meta;

// create a template
const Template: Story<InfiniteScrollListViewProps> = (args) => {
    return <InfiniteScrollListView {...args} />;
};

/** *************************************************
 * Data
 ************************************************** */

const sampleArgs: InfiniteScrollListViewProps = {
    onItemClick: (item: ListViewItem) => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/restrict-template-expressions
        alert(`${item.title} Clicked`);
    },
    dataToItemMapping: {
        id: 'id',
        title: 'title',
        message: 'message',
        displayDateTime: 'displayDateTime',
    },
    dataApiUrl: '/api/infinitescrolllistview/fakedata',
    dataLimit: 4,
    dataOffset: 0,
    dataOffsetIncrement: 4,
    height: '400px',
    onDeleteBtnClick: undefined, // have to manually undefined it because of storybook actions
    onEditBtnClick: undefined, // have to manually undefined it because of storybook actions
};

/** *************************************************
 * Stories
 ************************************************** */

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const Basic: Story = Template.bind({});
Basic.args = { ...sampleArgs };

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const EmptyList: Story = Template.bind({});
EmptyList.args = { ...Basic.args, dataApiUrl: '/api/emptydata' };

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const InvalidAPIEndPoint: Story = Template.bind({});
InvalidAPIEndPoint.args = { ...Basic.args, dataApiUrl: '/api/invalid' };

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const ListWithFlag: Story = Template.bind({});
ListWithFlag.args = {
    ...sampleArgs,
    dataToItemMapping: {
        id: 'id',
        title: 'title',
        message: 'message',
        displayDateTime: 'displayDateTime',
        isFlagged: 'isFlagged',
    },
};

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
export const ListWithControls: Story = Template.bind({});
ListWithControls.args = {
    ...sampleArgs,
    dataToItemMapping: {
        id: 'id',
        title: 'title',
        message: 'message',
        displayDateTime: 'displayDateTime',
    },
    onDeleteBtnClick: (item: ListViewItem): void => {
        console.log(`${item.title} is deleted`);
    },
    onEditBtnClick: (item: ListViewItem): void => {
        console.log(`${item.title} is edited`);
    },
};
