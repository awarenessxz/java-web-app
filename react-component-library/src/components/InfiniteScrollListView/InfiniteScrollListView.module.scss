.islvRoot {
  overflow-y: scroll;
  width: 100%;
  max-width: 36ch;
  background-color: #fafbfd;
  text-align: center;
  border: 1px solid #000000;
  padding: 1px;
}
