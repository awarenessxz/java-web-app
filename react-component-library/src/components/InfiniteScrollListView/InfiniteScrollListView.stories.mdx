import { Meta, Story, ArgsTable } from '@storybook/addon-docs/blocks';
import InfiniteScrollListView from "./InfiniteScrollListView";

<Meta title='Docs/InfiniteScrollListView' component={InfiniteScrollListView} />

# Infinite Scrolling List View Component

This is a infinite scrolling list view that is meant to mimic how outlook email list. This project is made possible
because of [joshtronic](https://www.digitalocean.com/community/tutorials/react-react-infinite-scroll). Check out his
explanation on how to implement infinite scrolling in React Component.

**Future Improvement**
- currently new data is appended into the list inside the component. If there is a change in order, this will break the
component. Hence, improvement will have to be made on data that changes dynamically.
- Read/Unread Feature
- Badges
- Write Test Cases!!

## Usage

To use this component, do the following:

1. Add the library as dependency
    - `yarn add react-component-library`
2. Import the component
    ```
    import { InfiniteScrollListView } from 'react-component-library';
    ```

## Properties

<ArgsTable of={InfiniteScrollListView} />

### Data API

The component will be dynamically calling a rest api to provide it with data. Below are some of the format you should adhere to:

#### 1. API end point

You should expose an API end point in the following format `/any/api/url?limit=?&offset=?`. The component assumes that
the end point is in the format above. Hence, a clearer picture of what the props in the properties table will be as
follows:
- dataApiUrl -- api url `eg. /any/api/url`
- dataLimit -- amount of data to fetch (Refer to SQL)
- dataOffset -- current position of data (Refer to SQL)

#### 2. Data to Item Mapping
This components accepts data in the following format:

```
{
    id: string;
    title: string;
    message: string;
}
```

Hence, we provided an `dataToItemMapping` prop for you to map the data which your api returns to the format accepted by
the component. Below is an example scenario.

- Fetch API returns the following object
    ```
    {
        id: 123123
        header: something,
        result: asdasdasd
    }
    ```

- When using the Component, set the `dataToItemMapping` parameter as shown
    ```
    <InfiniteScrollListView
        ...
        dataToItemMapping: {
            id: 'id',                   # maps id to id
            title: 'header',            # maps header to title
            message: 'result',          # maps result to message
        },
        ...
    />


