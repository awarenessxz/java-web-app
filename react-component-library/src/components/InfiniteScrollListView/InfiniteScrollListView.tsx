import React, { Fragment, useEffect, useState, useRef } from 'react';
import debounce from 'lodash.debounce';
import { createStyles, Theme, makeStyles } from '@material-ui/core/styles';
import Alert from '@material-ui/lab/Alert';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import Divider from '@material-ui/core/Divider';
import { InfiniteScrollListViewProps, InfiniteScrollListViewState, ListViewItem } from './InfiniteScrollListView.types';
import { getDateFromDateTime } from '../../utils/time-formatter';
import styles from './InfiniteScrollListView.module.scss';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        listContent: {
            width: '100%',
            backgroundColor: theme.palette.background.paper,
        },
        listItem: {
            cursor: 'pointer',
        },
        listIcon: {
            paddingTop: '10px',
        },
        listRightContent: {
            textAlign: 'right',
            flex: '0 0 60px',
            paddingTop: '10px',
        },
        inline: {
            display: 'inline',
        },
    }),
);

/**
 * Infinite Scroll List View
 */
const InfiniteScrollListView = (props: InfiniteScrollListViewProps): JSX.Element => {
    const classes = useStyles();
    const [localState, setLocalState] = useState<InfiniteScrollListViewState>({
        error: false,
        hasNewData: false,
        isLoading: true,
        dataOffset: props.dataOffset,
        data: [],
        triggerScrollRegister: false,
    });
    const [isDataFlagReadList, setIsDataFlagReadList] = useState<string[]>(props.dataIsFlaggedReadList || []);
    const componentRef = useRef<HTMLDivElement>(null);

    // https://www.geeksforgeeks.org/check-whether-html-element-has-scrollbars-using-javascript/
    const componentHasScrollbar = (): boolean => {
        const div = componentRef.current;
        let result = false;
        if (div !== null) {
            result = !!div.scrollTop;
            if (!result) {
                div.scrollTop = 1;
                result = !!div.scrollTop;
                div.scrollTop = 0;
            }
        }
        return result;
    };

    const loadData = debounce(() => {
        setLocalState({ ...localState, hasNewData: false, isLoading: true });
    }, 100);

    useEffect(() => {
        if (localState.isLoading) {
            const url = `${props.dataApiUrl}?limit=${props.dataLimit}&offset=${localState.dataOffset}`;
            fetch(url)
                .then((res) => res.json())
                .then((newData) => {
                    // map the data to item
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
                    const nextData: ListViewItem[] = newData.map((item: any) => {
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                        return {
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
                            id: item[props.dataToItemMapping.id],
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
                            title: item[props.dataToItemMapping.title],
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
                            message: item[props.dataToItemMapping.message],
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-assignment
                            displayDateTime: getDateFromDateTime(item[props.dataToItemMapping.displayDateTime]),
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                            isFlagged: props.dataToItemMapping.isFlagged
                                ? // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                                  item[props.dataToItemMapping.isFlagged as string]
                                : false,
                            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                            originalData: item,
                        };
                    });

                    // Merges the next data into our existing data
                    setLocalState({
                        ...localState,
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                        hasNewData: nextData.length > 0,
                        isLoading: false,
                        dataOffset: localState.dataOffset + props.dataOffsetIncrement,
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                        data: [...localState.data, ...nextData],
                        triggerScrollRegister: !localState.triggerScrollRegister,
                    });
                })
                .catch((err) => {
                    setLocalState({
                        ...localState,
                        isLoading: false,
                        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
                        error: 'Fail to load data',
                        // error: err.message,
                    });
                });
        } else if (!componentHasScrollbar() && localState.hasNewData) {
            // if there are space, but no scrollbar, and new data have been loaded
            // no new data --> means api return us empty []. Two possible scenarios:
            //      1. something might have went wrong, so don't load more data
            //      2. api has nothing to return
            // no scrollbar --> means div is too large, try to load more data to fill it, so that scrollbar appears
            loadData();
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [localState]);

    const handleOnScroll = (e: Event): void => {
        // Bails early if:
        // * there's an error
        // * it's already loading
        // * there's nothing left to load
        if (localState.error || localState.isLoading || !localState.hasNewData) return;

        // Checks that the page has scrolled to the bottom
        const target = e.target as HTMLDivElement;
        if (target.scrollHeight - target.scrollTop === target.clientHeight) {
            loadData();
        }
    };

    // useEffect to register scroll listener, will be trigger whenever state triggerScrollRegister is updated
    useEffect(() => {
        const div = componentRef.current;
        // subscribe event
        if (div != null) {
            // console.log('registering scroll listener');
            div.addEventListener('scroll', handleOnScroll);
        }
        return (): void => {
            // unsubscribe event
            if (div != null) {
                // console.log('unregistering scroll listener');
                div.removeEventListener('scroll', handleOnScroll);
            }
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [localState.triggerScrollRegister]);

    const ListContentItem = (item: ListViewItem): JSX.Element => {
        return (
            <Fragment>
                <ListItem
                    alignItems="flex-start"
                    onClick={(): void => {
                        // update isDataReadList
                        setIsDataFlagReadList([...isDataFlagReadList, item.id]);
                        props.onItemClick(item.originalData);
                    }}
                    className={`${classes.listItem} ${
                        item.isFlagged && !isDataFlagReadList.includes(item.id)
                            ? styles.islvIsFlagged
                            : styles.islvIsNotFlagged
                    }`}
                    data-testid={`islv_${item.id}`}
                >
                    <ListItemText primary={item.title} secondary={item.message} />
                    <ListItemText secondary={item.displayDateTime} className={classes.listRightContent} />
                </ListItem>
                <Divider component="li" />
            </Fragment>
        );
    };

    const ListContent = (): JSX.Element => {
        return (
            <Fragment>
                {localState.data.length > 0 ? (
                    <List className={classes.listContent}>
                        {localState.data.map((item, idx) => (
                            <ListContentItem key={idx} {...item} />
                        ))}
                    </List>
                ) : (
                    <Alert data-testid="islv_emptyDiv" severity="success">
                        No Data Available
                    </Alert>
                )}
                {localState.isLoading && <div data-testid="islv_isLoadingDiv">Loading...</div>}
                {!localState.hasNewData && (
                    <Alert data-testid="islv_endDiv" severity="info">
                        --- End of List ---
                    </Alert>
                )}
            </Fragment>
        );
    };

    return (
        <div
            data-testid="islv_rootDiv"
            className={styles.islvRoot}
            style={{ height: props.height, width: props.width }}
            ref={componentRef}
        >
            {localState.error ? (
                <Alert data-testid="islv_errorDiv" severity="error">
                    {localState.error}
                </Alert>
            ) : (
                <ListContent />
            )}
        </div>
    );
};

export default InfiniteScrollListView;
