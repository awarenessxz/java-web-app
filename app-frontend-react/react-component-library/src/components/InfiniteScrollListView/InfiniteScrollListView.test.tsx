import * as React from 'react';
import 'whatwg-fetch';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { render, RenderResult, screen, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';

// Import Component
import InfiniteScrollListView from './InfiniteScrollListView';
import { InfiniteScrollListViewProps } from './InfiniteScrollListView.types';
import { getFakeData } from './InfiniteScrollListView.util';

// set up mock server for rest api end point
const server = setupServer(
    rest.get('/test/infinitescrolllistview/fakedata', (req, res, ctx) => {
        const data = getFakeData(1, 0);
        return res(ctx.json(data));
    }),
    rest.get('/test/infinitescrolllistview/empty', (req, res, ctx) => {
        return res(ctx.json([]));
    }),
);
beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// function to render Component before each test
const renderComponent = (props = {}): RenderResult => {
    const defaultProps: InfiniteScrollListViewProps = {
        onItemClick: (item) => jest.fn(),
        dataToItemMapping: {
            id: 'id',
            title: 'title',
            message: 'message',
            displayDateTime: 'displayDateTime',
        },
        dataApiUrl: '/test/infinitescrolllistview/fakedata',
        dataLimit: 4,
        dataOffset: 0,
        dataOffsetIncrement: 4,
        height: '400px',
    };
    const merged = { ...defaultProps, ...props };
    return render(<InfiniteScrollListView {...merged} />);
};

// 1. Testing if component renders properly
describe('Testing if component renders properly', () => {
    // Snapshot Testing
    it('Snapshot Testing', async () => {
        const { asFragment } = renderComponent();
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        expect(asFragment()).toMatchSnapshot();
    });
});

// 2. Test API
describe('Testing API Fetch Logic', () => {
    // fail to load data as end point can't be hit
    it('Invalid API', async () => {
        console.error = jest.fn(); // disable console.error
        renderComponent({ dataApiUrl: '/test/infinitescrolllistview/invalid' });
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        expect(screen.getByTestId('islv_errorDiv')).toBeInTheDocument();
    });

    // api returns empty list
    it('Empty List', async () => {
        renderComponent({ dataApiUrl: '/test/infinitescrolllistview/empty' });
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        expect(screen.getByTestId('islv_endDiv')).toBeInTheDocument();
    });

    // api returns result
    it('default with results', async () => {
        // only renders the first data in the list
        renderComponent({ dataApiUrl: '/test/infinitescrolllistview/fakedata', dataLimit: 1, dataOffset: 0 });
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        expect(screen.getByText('Alan1')).toBeInTheDocument();
    });
});

// 3. Event Listener
describe('Testing OnClick Events', () => {
    it('Event triggered when List Item is clicked', async () => {
        const mockClickHandler = jest.fn();
        renderComponent({
            dataApiUrl: '/test/infinitescrolllistview/fakedata',
            dataLimit: 1,
            dataOffset: 0,
            onItemClick: mockClickHandler,
        });
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        const listItem1 = screen.getByTestId('islv_1');
        fireEvent.click(listItem1);
        expect(mockClickHandler).toBeCalled();
    });
});

// 4. Test Scrolling -- NOT WORKING, seems to have limitation for such testing
// https://github.com/testing-library/react-testing-library/issues/671
// https://github.com/testing-library/react-testing-library/issues/353
describe('Testing API Fetch Logic', () => {
    /*
    const originalOffsetHeight = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetHeight');
    const originalOffsetWidth = Object.getOwnPropertyDescriptor(HTMLElement.prototype, 'offsetWidth');

    beforeEach(() => {
        Object.defineProperty(HTMLElement.prototype, 'offsetHeight', { configurable: true, value: 2000 });
        Object.defineProperty(HTMLElement.prototype, 'offsetWidth', { configurable: true, value: 2000 });
    });

    afterEach(() => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        Object.defineProperty(HTMLElement.prototype, 'offsetHeight', originalOffsetHeight);
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        Object.defineProperty(HTMLElement.prototype, 'offsetWidth', originalOffsetWidth);
    });

    it('scroll for more data', async () => {
        renderComponent({ dataApiUrl: '/test/infinitescrolllistview/fakedata', dataLimit: 2, dataOffset: 0 });
        await waitForElementToBeRemoved(screen.getByTestId('islv_isLoadingDiv'));
        const div = screen.getByTestId('islv_rootDiv');
        console.log(div.clientHeight, div.scrollHeight, div.scrollTop);
        div.clientHeight = 500;
        fireEvent.scroll(div, { target: { scrollY: 800 } });
        console.log(div.clientHeight, div.scrollHeight, div.scrollTop);
        screen.debug();
    });
     */
});
