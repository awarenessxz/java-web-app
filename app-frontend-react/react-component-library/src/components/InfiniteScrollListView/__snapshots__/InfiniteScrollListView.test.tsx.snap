// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if component renders properly Snapshot Testing 1`] = `
<DocumentFragment>
  <div
    class="islvRoot"
    data-testid="islv_rootDiv"
    style="height: 400px;"
  >
    <ul
      class="MuiList-root makeStyles-listContent-1 MuiList-padding"
    >
      <li
        class="MuiListItem-root makeStyles-listItem-2 islvIsNotFlagged MuiListItem-gutters MuiListItem-alignItemsFlexStart"
        data-testid="islv_1"
      >
        <div
          class="MuiListItemText-root MuiListItemText-multiline"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            Alan1
          </span>
          <p
            class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            this is just a previewthis is just a previewthis is just a preview this is just a preview this is just a preview this is just a preview this is just a preview
          </p>
        </div>
        <div
          class="MuiListItemText-root makeStyles-listRightContent-4"
        >
          <p
            class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
          >
            04/12/2020
          </p>
        </div>
      </li>
      <li
        class="MuiDivider-root"
        role="separator"
      />
    </ul>
  </div>
</DocumentFragment>
`;
